### PR TITLE
fix(moderation): increase duration lengths

### PR DESCRIPTION
**Pull request by kyranet**
Executed query:
```sql
query: START TRANSACTION
query: SELECT * FROM current_schema()
query: SELECT * FROM "information_schema"."tables" WHERE ("table_schema" = 'public' AND "table_name" = 'guilds')
query: SELECT *, ('"' || "udt_schema" || '"."' || "udt_name" || '"')::"regtype" AS "regtype" FROM "information_schema"."columns" WHERE ("table_schema" = 'public' AND "table_name" = 'guilds')
query: SELECT "ns"."nspname" AS "table_schema", "t"."relname" AS "table_name", "cnst"."conname" AS "constraint_name", pg_get_constraintdef("cnst"."oid") AS "expression", CASE "cnst"."contype" WHEN 'p' THEN 'PRIMARY' WHEN 'u' THEN 'UNIQUE' WHEN 'c' THEN 'CHECK' WHEN 'x' THEN 'EXCLUDE' END AS "constraint_type", "a"."attname" AS "column_name" FROM "pg_constraint" "cnst" INNER JOIN "pg_class" "t" ON "t"."oid" = "cnst"."conrelid" INNER JOIN "pg_namespace" "ns" ON "ns"."oid" = "cnst"."connamespace" LEFT JOIN "pg_attribute" "a" ON "a"."attrelid" = "cnst"."conrelid" AND "a"."attnum" = ANY ("cnst"."conkey") WHERE "t"."relkind" = 'r' AND (("ns"."nspname" = 'public' AND "t"."relname" = 'guilds'))
query: SELECT "ns"."nspname" AS "table_schema", "t"."relname" AS "table_name", "i"."relname" AS "constraint_name", "a"."attname" AS "column_name", CASE "ix"."indisunique" WHEN 't' THEN 'TRUE' ELSE'FALSE' END AS "is_unique", pg_get_expr("ix"."indpred", "ix"."indrelid") AS "condition", "types"."typname" AS "type_name" FROM "pg_class" "t" INNER JOIN "pg_index" "ix" ON "ix"."indrelid" = "t"."oid" INNER JOIN "pg_attribute" "a" ON "a"."attrelid" = "t"."oid"  AND "a"."attnum" = ANY ("ix"."indkey") INNER JOIN "pg_namespace" "ns" ON "ns"."oid" = "t"."relnamespace" INNER JOIN "pg_class" "i" ON "i"."oid" = "ix"."indexrelid" INNER JOIN "pg_type" "types" ON "types"."oid" = "a"."atttypid" LEFT JOIN "pg_constraint" "cnst" ON "cnst"."conname" = "i"."relname" WHERE "t"."relkind" = 'r' AND "cnst"."contype" IS NULL AND (("ns"."nspname" = 'public' AND "t"."relname" = 'guilds'))
query: SELECT "con"."conname" AS "constraint_name", "con"."nspname" AS "table_schema", "con"."relname" AS "table_name", "att2"."attname" AS "column_name", "ns"."nspname" AS "referenced_table_schema", "cl"."relname" AS "referenced_table_name", "att"."attname" AS "referenced_column_name", "con"."confdeltype" AS "on_delete", "con"."confupdtype" AS "on_update", "con"."condeferrable" AS "deferrable", "con"."condeferred" AS "deferred" FROM ( SELECT UNNEST ("con1"."conkey") AS "parent", UNNEST ("con1"."confkey") AS "child", "con1"."confrelid", "con1"."conrelid", "con1"."conname", "con1"."contype", "ns"."nspname", "cl"."relname", "con1"."condeferrable", CASE WHEN "con1"."condeferred" THEN 'INITIALLY DEFERRED' ELSE 'INITIALLY IMMEDIATE' END as condeferred, CASE "con1"."confdeltype" WHEN 'a' THEN 'NO ACTION' WHEN 'r' THEN 'RESTRICT' WHEN 'c' THEN 'CASCADE' WHEN 'n' THEN 'SET NULL' WHEN 'd' THEN 'SET DEFAULT' END as "confdeltype", CASE "con1"."confupdtype" WHEN 'a' THEN 'NO ACTION' WHEN 'r' THEN 'RESTRICT' WHEN 'c' THEN 'CASCADE' WHEN 'n' THEN 'SET NULL' WHEN 'd' THEN 'SET DEFAULT' END as "confupdtype" FROM "pg_class" "cl" INNER JOIN "pg_namespace" "ns" ON "cl"."relnamespace" = "ns"."oid" INNER JOIN "pg_constraint" "con1" ON "con1"."conrelid" = "cl"."oid" WHERE "con1"."contype" = 'f' AND (("ns"."nspname" = 'public' AND "cl"."relname" = 'guilds')) ) "con" INNER JOIN "pg_attribute" "att" ON "att"."attrelid" = "con"."confrelid" AND "att"."attnum" = "con"."child" INNER JOIN "pg_class" "cl" ON "cl"."oid" = "con"."confrelid" INNER JOIN "pg_namespace" "ns" ON "cl"."relnamespace" = "ns"."oid" INNER JOIN "pg_attribute" "att2" ON "att2"."attrelid" = "con"."conrelid" AND "att2"."attnum" = "con"."parent"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_0c926b1228ef524fb163860aa6"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_6549fcc4eaf52aaa25b1434efd"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_d8912ecae85bd3a46b898f0690"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_ef4cc47769236585d2b3f4a37a"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_1deb783659bfac3ff4b0e40f19"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_4df15ef3861b2e32f5fba93a1e"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_3efb778a340d97d722840cbe98"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_59399065ad22d910f75b3de434"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_1607b475d70211cf9af1a20d5e"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_b0c8ac0c023bfee0e21a58dc66"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_0e3795e2cdf5d64ac1ac78a405"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_e031b9117a33419f33a84af19c"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_8083d38a71e559b88f0ff9af60"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_2540e88ea8b07982f481874b3d"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_36f92b62f8537c824b90525c07"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_7688251bcbd5be20e6c9e358c1"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_b15f81e1bf62afe7f993a95815"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_f1f90a5eb6e257e6b307091fc2"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_9d22cf0eb1218598850f75feef"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_681edc2018e782817478b731f5"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_067d4759b039d3f11bf9790dd3"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_5d04b1b63a1a4fbc0ac7543b61"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_88e12d03ecf6993214bbe2d67d"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_1fad7b108fbd2218dba944214c"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_6d2034879144b6e06b351ccf23"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_42e3b8013d63eefe00bdc793dd"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_c2f47d7253e32917b0d028c85d"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_b0528ca04c553d9d0e4b4fafcc"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_711487c0dcac3084b186d4d3a4"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_2f496efbf878214b66c69d4a9b"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_6976c3ea0b1910c54c78defb8b"
query: ALTER TABLE "guilds" DROP CONSTRAINT "CHK_c732b8ad57c50528ac69d12eb2"
query: ALTER TABLE public.guilds ALTER "selfmod.attachments.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.capitals.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.links.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.messages.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.newlines.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.invites.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.filter.hardActionDuration" TYPE bigint
query: ALTER TABLE public.guilds ALTER "selfmod.reactions.hardActionDuration" TYPE bigint
query: SELECT * FROM current_schema()
query: SELECT * FROM "information_schema"."tables" WHERE ("table_schema" = 'public' AND "table_name" = 'moderation')
query: SELECT *, ('"' || "udt_schema" || '"."' || "udt_name" || '"')::"regtype" AS "regtype" FROM "information_schema"."columns" WHERE ("table_schema" = 'public' AND "table_name" = 'moderation')
query: SELECT "ns"."nspname" AS "table_schema", "t"."relname" AS "table_name", "cnst"."conname" AS "constraint_name", pg_get_constraintdef("cnst"."oid") AS "expression", CASE "cnst"."contype" WHEN 'p' THEN 'PRIMARY' WHEN 'u' THEN 'UNIQUE' WHEN 'c' THEN 'CHECK' WHEN 'x' THEN 'EXCLUDE' END AS "constraint_type", "a"."attname" AS "column_name" FROM "pg_constraint" "cnst" INNER JOIN "pg_class" "t" ON "t"."oid" = "cnst"."conrelid" INNER JOIN "pg_namespace" "ns" ON "ns"."oid" = "cnst"."connamespace" LEFT JOIN "pg_attribute" "a" ON "a"."attrelid" = "cnst"."conrelid" AND "a"."attnum" = ANY ("cnst"."conkey") WHERE "t"."relkind" = 'r' AND (("ns"."nspname" = 'public' AND "t"."relname" = 'moderation'))
query: SELECT "ns"."nspname" AS "table_schema", "t"."relname" AS "table_name", "i"."relname" AS "constraint_name", "a"."attname" AS "column_name", CASE "ix"."indisunique" WHEN 't' THEN 'TRUE' ELSE'FALSE' END AS "is_unique", pg_get_expr("ix"."indpred", "ix"."indrelid") AS "condition", "types"."typname" AS "type_name" FROM "pg_class" "t" INNER JOIN "pg_index" "ix" ON "ix"."indrelid" = "t"."oid" INNER JOIN "pg_attribute" "a" ON "a"."attrelid" = "t"."oid"  AND "a"."attnum" = ANY ("ix"."indkey") INNER JOIN "pg_namespace" "ns" ON "ns"."oid" = "t"."relnamespace" INNER JOIN "pg_class" "i" ON "i"."oid" = "ix"."indexrelid" INNER JOIN "pg_type" "types" ON "types"."oid" = "a"."atttypid" LEFT JOIN "pg_constraint" "cnst" ON "cnst"."conname" = "i"."relname" WHERE "t"."relkind" = 'r' AND "cnst"."contype" IS NULL AND (("ns"."nspname" = 'public' AND "t"."relname" = 'moderation'))
query: SELECT "con"."conname" AS "constraint_name", "con"."nspname" AS "table_schema", "con"."relname" AS "table_name", "att2"."attname" AS "column_name", "ns"."nspname" AS "referenced_table_schema", "cl"."relname" AS "referenced_table_name", "att"."attname" AS "referenced_column_name", "con"."confdeltype" AS "on_delete", "con"."confupdtype" AS "on_update", "con"."condeferrable" AS "deferrable", "con"."condeferred" AS "deferred" FROM ( SELECT UNNEST ("con1"."conkey") AS "parent", UNNEST ("con1"."confkey") AS "child", "con1"."confrelid", "con1"."conrelid", "con1"."conname", "con1"."contype", "ns"."nspname", "cl"."relname", "con1"."condeferrable", CASE WHEN "con1"."condeferred" THEN 'INITIALLY DEFERRED' ELSE 'INITIALLY IMMEDIATE' END as condeferred, CASE "con1"."confdeltype" WHEN 'a' THEN 'NO ACTION' WHEN 'r' THEN 'RESTRICT' WHEN 'c' THEN 'CASCADE' WHEN 'n' THEN 'SET NULL' WHEN 'd' THEN 'SET DEFAULT' END as "confdeltype", CASE "con1"."confupdtype" WHEN 'a' THEN 'NO ACTION' WHEN 'r' THEN 'RESTRICT' WHEN 'c' THEN 'CASCADE' WHEN 'n' THEN 'SET NULL' WHEN 'd' THEN 'SET DEFAULT' END as "confupdtype" FROM "pg_class" "cl" INNER JOIN "pg_namespace" "ns" ON "cl"."relnamespace" = "ns"."oid" INNER JOIN "pg_constraint" "con1" ON "con1"."conrelid" = "cl"."oid" WHERE "con1"."contype" = 'f' AND (("ns"."nspname" = 'public' AND "cl"."relname" = 'moderation')) ) "con" INNER JOIN "pg_attribute" "att" ON "att"."attrelid" = "con"."confrelid" AND "att"."attnum" = "con"."child" INNER JOIN "pg_class" "cl" ON "cl"."oid" = "con"."confrelid" INNER JOIN "pg_namespace" "ns" ON "cl"."relnamespace" = "ns"."oid" INNER JOIN "pg_attribute" "att2" ON "att2"."attrelid" = "con"."conrelid" AND "att2"."attnum" = "con"."parent"
query: ALTER TABLE public.moderation ALTER "duration" TYPE bigint
query: ALTER TABLE "moderation" DROP CONSTRAINT "CHK_c264fea544708ca3f65d1d027c"
query: COMMIT
```

Data is also read fine as numbers:
image

And moderation cases with a length higher than 24 days 18 hours also work, testing one year:
image

### CODE REVIEW

**Review by vladfrangu — CHANGES_REQUESTED** [review]
This feels like a literal... "hey let's drop the limit and thats all", instead of a "hey lets support BigInts, literally"... Unless you change the types too to BigInt.. idk? LGTM?

**Comment by kyranet** [comment]
> This feels like a literal... "hey let's drop the limit and thats all", instead of a "hey lets support BigInts, literally"... Unless you change the types too to BigInt.. idk? LGTM?

`integer` limits the duration of the moderation cases to roughly 24 days 18 hours, and there are many cases where mods want to apply an action that lasts for months. `bigint` is suitable for this.

As for ranges, we check that in the commands.

**Comment by vladfrangu** [comment]
I'm talking about the JS type bigint, not the db one. Those values are still interpreted as a number in JS, and stored as a String on the DB (which I assume then gets converted to a bigint by the DB)

**Comment by kyranet** [comment]
Those values are interpreted as a string, and sent as a string. The transformer deserializes the string into a number, and the number into a string.

File: src/lib/database/entities/GuildEntity.ts [hunk 1]
```diff
@@ -5,47 +5,16 @@ import { AnyObject } from '#lib/types';
...
-@Check(/* sql */ `("selfmod.reactions.thresholdDuration" >= 0) AND ("selfmod.reactions.thresholdDuration" <= 120000)`)
-@Check(/* sql */ `"selfmod.messages.hardActionDuration" >= 1000`)
-@Check(/* sql */ `("selfmod.messages.thresholdMaximum" >= 0) AND ("selfmod.messages.thresholdMaximum" <= 60)`)
-@Check(/* sql */ `("selfmod.messages.thresholdDuration" >= 0) AND ("selfmod.messages.thresholdDuration" <= 120000)`)
-@Check(/* sql */ `"selfmod.links.hardActionDuration" >= 1000`)
-@Check(/* sql */ `("selfmod.links.thresholdMaximum" >= 0) AND ("selfmod.links.thresholdMaximum" <= 60)`)
-@Check(/* sql */ `("selfmod.links.thresholdDuration" >= 0) AND ("selfmod.links.thresholdDuration" <= 120000)`)
-@Check(/* sql */ `("selfmod.raidthreshold" >= 2) AND ("selfmod.raidthreshold" <= 50)`)
-@Check(/* sql */ `"no-mention-spam.mentionsAllowed" >= 0`)
-@Check(/* sql */ `"no-mention-spam.timePeriod" >= 0`)
```
Comment (favna) [inline src/lib/database/entities/GuildEntity.ts:47]: These are not being re-added on the entity fields

File: src/lib/database/entities/ModerationEntity.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@ export class ModerationEntity extends BaseEntity {
 	@Column('timestamp without time zone', { nullable: true, default: () => 'null' })
 	public createdAt: Date | null = null;
 
-	@Column('integer', { nullable: true, default: () => 'null' })
```
Comment (favna) [inline src/lib/database/entities/ModerationEntity.ts:31]: add back minimum and maximum

**Review by favna — CHANGES_REQUESTED** [review]
I could go on pasting the image a lot more times but you get the picture.

Please re-add minimum and maximum properly.

File: src/lib/database/entities/GuildEntity.ts [hunk 3]
```diff
@@ -310,15 +279,15 @@ export class GuildEntity extends BaseEntity {
 	@Column('smallint', { name: 'selfmod.attachments.hardAction', default: 0 })
 	public selfmodAttachmentsHardAction = 0;
 
-	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey })
-	@Column('integer', { name: 'selfmod.attachments.hardActionDuration', nullable: true })
+	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey, minimum: Time.Second, maximum: Time.Year * 5 })
+	@Column('bigint', { name: 'selfmod.attachments.hardActionDuration', nullable: true, transformer: kBigIntTransformer })
 	public selfmodAttachmentsHardActionDuration: number | null = null;
 
 	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey })
```
Comment (favna) [inline src/lib/database/entities/GuildEntity.ts]: re-add minimum and maximum from checks

File: src/lib/database/entities/GuildEntity.ts [hunk 4]
```diff
@@ -462,15 +431,15 @@ export class GuildEntity extends BaseEntity {
...
+	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey, minimum: Time.Second, maximum: Time.Year * 5 })
+	@Column('bigint', { name: 'selfmod.newlines.hardActionDuration', nullable: true, transformer: kBigIntTransformer })
 	public selfmodNewlinesHardActionDuration: number | null = null;
 
 	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey })
 	@Column('smallint', { name: 'selfmod.newlines.thresholdMaximum', default: 10 })
 	public selfmodNewlinesThresholdMaximum = 10;
 
-	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey })
+	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey, minimum: 0, maximum: Time.Minute * 5 })
```
Comment (favna) [inline src/lib/database/entities/GuildEntity.ts:442]: 5 minutes of milliseconds is not the same as the old maximum value of `120000`

File: src/lib/database/entities/GuildEntity.ts [hunk 5]
```diff
@@ -310,15 +279,15 @@ export class GuildEntity extends BaseEntity {
 	@Column('smallint', { name: 'selfmod.attachments.hardAction', default: 0 })
 	public selfmodAttachmentsHardAction = 0;
 
-	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey })
-	@Column('integer', { name: 'selfmod.attachments.hardActionDuration', nullable: true })
+	@ConfigurableKey({ dashboardOnly: true, description: LanguageKeys.Settings.DashboardOnlyKey, minimum: Time.Second, maximum: Time.Year * 5 })
```
Comment (favna) [inline src/lib/database/entities/GuildEntity.ts:282]: Check didn't have a minimum
And `Time.Year * 5` is not the same as `1000`

**Review by favna — CHANGES_REQUESTED** [review]

File: src/lib/database/entities/GuildEntity.ts (hunk 5 above)
Comment (kyranet) [reply src/lib/database/entities/GuildEntity.ts:282]:
```sql
@Check(/* sql */ `"selfmod.attachments.hardActionDuration" >= 1000`)
```

Here the minimum is set to Time.Second, which is `1000`.

File: src/lib/database/entities/GuildEntity.ts (hunk 4 above)
Comment (kyranet) [reply src/lib/database/entities/GuildEntity.ts:442]: It's an increase, 2 minutes is too low for many cases.

Once we move adders to Redis, they'll increase to 10 or 15 minutes.

**Review by favna — APPROVED** [review]

**Pull request merged**